runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java

```diff
@@ -1264,6 +1268,10 @@ private static void translateFn(
     // in streaming but does not work in batch
     if (context.getPipelineOptions().isStreaming() && isStateful) {
       stepContext.addInput(PropertyNames.USES_KEYED_STATE, "true");
```

Comment:
Is `ALLOWS_SHARDABLE_STATE ` supported by both streaming engine and windmill appliance? If it's only for streaming engine, is it safe to populate this field for both?

Reply:
As a first step, it's only supported by streaming engine. But it's fine to set it in both cases since the field will be ignored in appliance. We will need to wait for the backend to recognize this property anyway, otherwise the pipeline will check fail.

Reply:
Update on this:

I added an experiment to gate the auto-sharding so this can be merged without waiting for the backend. It will also make the testing easier.

I also added a check for the experiment, "beam_fn_api". My intention was to disable the feature for unified worker but I guess this way we would disable auto-sharding for both unified worker and java worker using fn api - I remember that we are not going to support the latter so it seems fine to me. But let me know if my understanding is incorrect.